src/GSMConnectionManager.h

```diff
@@ -26,50 +25,171 @@ class GSMConnectionManager : public ConnectionManager {
+        debugMessage("Connection to Cellular Network lost during Time acquisition.\nAttempting reconnection", 0);
+      }
+      newInterval = CHECK_INTERVAL_DISCONNECTED;
+      break;
+    case CONNECTION_STATE_ERROR:
+      debugMessage("GPRS attach failed\nMake sure the antenna is connected and reset your board.", 0);
+      break;
   }
+  connectionTickTimeInterval = newInterval;
+  lastConnectionTickTime = millis();
+  netConnectionState = _newState;
+}
 
-  Serial.print("<<Network Status: ");
-  if (gprs.attachGPRS(apn, login, pass) == GPRS_READY) {
-    Serial.println("CONNECTED");
-    lastNetworkCheck = millis();
-  } else {
-    Serial.println("NOT CONNECTED");
+void GSMConnectionManager::check() {
+  char msgBuffer[120];
+  unsigned long const now = millis();
+  int gsmAccessAlive;
+  if (now - lastConnectionTickTime > connectionTickTimeInterval) {
+    switch (netConnectionState) {
+      case CONNECTION_STATE_INIT:
+        init();
+        break;
+      case CONNECTION_STATE_CONNECTING:
+        /***  Blocking Call when 4th parameter == true   ***/
+        GSM3_NetworkStatus_t networkStatus;
+        networkStatus = gprs.attachGPRS(apn, login, pass, true);
+        sprintf(msgBuffer, "GPRS.attachGPRS(): %d", networkStatus);
+        debugMessage(msgBuffer, 3);
+        if (networkStatus == GSM3_NetworkStatus_t::ERROR) {
+          // NO FURTHER ACTION WILL FOLLOW THIS
+          changeConnectionState(CONNECTION_STATE_ERROR);
+          return;
+        }
+        debugMessage("Sending PING to outer space...", 2);
+        int pingResult;
+        pingResult = gprs.ping("google.com");
+        sprintf(msgBuffer, "GSM.ping(): %d", pingResult);
+        debugMessage(msgBuffer, 2);
+        if (pingResult < 0) {
+          debugMessage("PING failed", 0);
+          sprintf(msgBuffer, "Retrying in  \"%d\" milliseconds", connectionTickTimeInterval);
+          debugMessage(msgBuffer, 2);
+          return;
+        } else {
+          sprintf(msgBuffer, "Connected to GPRS Network");
+          debugMessage(msgBuffer, 2);
+          changeConnectionState(CONNECTION_STATE_GETTIME);
+          return;
+        }
+        break;
+      case CONNECTION_STATE_GETTIME:
+        debugMessage("Acquiring Time from Network", 3);
+        unsigned long networkTime;
+        networkTime = getTime();
+        debugMessage(".", 3, false, false);
+        if(networkTime > lastValidTimestamp){
+          lastValidTimestamp = networkTime;
+          sprintf(msgBuffer, "Network Time: %u", networkTime);
+          debugMessage(msgBuffer, 3);
+          changeConnectionState(CONNECTION_STATE_CONNECTED);
+        }else if(gsmAccess.isAccessAlive() != 1){
+          changeConnectionState(CONNECTION_STATE_DISCONNECTED);
+        }else if (!getTimeRetries--) {
+           changeConnectionState(CONNECTION_STATE_DISCONNECTED);
+        }
+        break;
+      case CONNECTION_STATE_CONNECTED:
+        /***  keep testing network connection   ***/
```

Comment:
thanks, Per!
this and the previous one (same style issue) are in my opinion more readable as I wrote them, and although they may disappear soon I'd like some suggestion from you on how to make them stand out even as a single line comment.
I think you object to the use of multi-line comment practices in lieu of a single line :)

Reply:
The problem with this style of comment is that it doesn't permit you to easily block comment out sections of code that contain them for testing or development (yes, I know you can use `#if false ... #endif` as an alternative). I received a specific complaint about this from a member of the Arduino community about Arduino code just yesterday, which was my primary motivation for this suggestion. I don't think the improvement in readability outweighs this disadvantage. It seems to me that it's the `***` that are really providing the readability, and you can still leave those in the single line comment if you like:
```c++
// *** keep testing network connection ***
```

I also don't like the inconsistency. The rest of the code uses single line comments and I'd like to see that style followed throughout the code. Style inconsistency is problematic for people who later want to add to the code because they need to waste time trying to figure out "should I use single line comment style, or block comment style?".

In the end, I don't have very strong feelings about this subject. If you prefer to ignore my suggestion on this subject, that's fine.

Reply:
hey @per1234 
makes a ton of sense, thank you for citing a case and reporting about the complaint you received.
Because these are still requiring attention for the near future but do not deserve the state of TODO I'll switch to single line comment and leave the `***`.
Inconsistency is bad, definitely. I'm glad you are taking the time to go through code readability and report anything odd that you find.
Life-saver, mate 👍